### DISCUSSION TITLE
How can I load more Java ruleset into Knowledge or its sessions?

### DISCUSSION THREAD

**Discussion by hailg-helm** [evrete/evrete#55] [General]
Hi, I want to build a system, where new rules (written in Java) can be loaded into a Knowledge or a stateful session in runtime. How can I achieve this? 

To be more specific, new rules will use the Java annotation to define @RuleSet and @Rule, and I want to load them into Knowledge/Session after they have been created.

Thank you.

**Comment by andbi** [comment]
Hi, could you please elaborate more on the "loaded in runtime"? What is wrong with the runtime when you obtain a `RuleBuilder` from a `Knowledge` or `StatefulSession` instance and start adding rules?

**Comment by hailg-helm** [comment]
For example, let's say I want to build a REST API service on top of this. My API could be sth like:

1. addNewRule(ruleUrl: String);
2. classifyData(object: Map<String, String>)

For the 1st API, the user can send me a ruleUrl, pointing to a Java source file. Then I want to fetch this URL, and the add the rule to the Knowledge/Session.

Thanks.

**Reply by hailg-helm** [reply]
To elaborate more, what I'm testing right now (and it's working well) is this

```
KnowledgeService service = new KnowledgeService();
Knowledge knowledge = service
        .newKnowledge(
                "JAVA-SOURCE", // <-- name of the DSL implementation
                new URL("https://www.evrete.org/examples/PrimeNumbersSource.java")
        );

StatefulSession session = knowledge.newStatefulSession();
```

What I need after getting this `knowledge`/`session` instance is to add more Java rules to them.

**Reply by andbi** [reply]
The good news is that the `newKnowledge` method actually accepts `URL` arrays, so your REST API seems to be possible to implement by accumulating urls and building the knowledge in a final call. 

The bad news is that combining  DSL knowledge sources isn't possible for sessions, there were impassable obstacles for that in versions before 3.0.0. I will need to take some time and see if they're still a stopper after the recent major update.

**Reply by hailg-helm** [reply]
Yeah, I've tried to pass a few URLs and it's working. But what I want is after the Knowledge is created, I cannot pass more Java sources into it, right? 

From all the documents I found so far, I can add a rule as a String expression, but no Java source.

**Reply by andbi** [reply]
Yes, among other smaller problems, this one boils down to finding a consistent way of combining two separate rulesets. There's no immediate solution here, but I will look into it.

**Reply by hailg-helm** [reply]
Thank you. :)
